### ISSUE TITLE
Replication not acknowledging latest WAL message

### ISSUE THREAD

**Issue by jeffutter** [epgsql/epgsql#179]
I am testing out a setup very similar to what is outlined in `streaming.md` to receive the WAL log. The main difference is I am using the `wal2json` logical decoder.

If I start listening, make some changes in the DB in a few transactions, and commit them. I see the changes dumped to my console. If I then kill my beam process and start it again, it replays the last message received. I can kill it any number of times, it always replays the last message.

Using `pg_recvlogical` from the command line I only see this issue if I kill `pg_recvlogical` quickly after receiving the message. But if I wait a few seconds it writes the `Standby status update` and when I next launch the command it doesn't display any messages until new ones come in.

I dumped the chatter between both `pg_recvlogical` and `epgsql` with wireshark. It looks like `pg_recvlogical` waits ~10 seconds to send the `Standby status update` while `epgsql` sends it immediately. I would imagine I could kill the beam process a few seconds after it received the message and the postgres server should be updated by that point.

The only thing I've noticed when combing through the code is that the postgres docs say the `Standby status update` should contain: 

```
Standby status update (F)
Byte1('r')
Identifies the message as a receiver status update.

Int64
The location of the last WAL byte + 1 received and written to disk in the standby.
```

The key being "location of the last WAL byte +1". I don't see the `+1` happening anywhere in the code. I did however manually `+1` the `endLsn` in my `handle_x_log_data` function and it didn't seem to make a difference.

Has anyone else experienced this, are my expectations about how this should work incorrect or am I missing anything obvious?

**Comment by seriyps** [comment]
Hi,
sorry for late reply! Missed your report while being on vacation.
Unfortunately I don't know much about replication protocol. Maybe @MikhailKalashnikov can suggest something?

**Comment by jarlopez** [comment]
I've been seeing this behavior with the following setup:
`wal2json` plugin, PG 9.6, asynchronous mode where the callback responds to `handle_cast` messages. I have a timer which handles the data and then sends the status update with the last received `end_lsn`. After receiving OK from `epgsql`, I `Ctrl+C` the application and restart it. After requesting `start_replication` with `"/0"` as the starting LSN, I see the last committed message again.

**Comment by jeffutter** [comment]
Sorry for taking so long to respond on this. I ended up not using wal replication on my project. However, I would like to eventually get it working.

My situation was similar to what was outlined here: https://github.com/epgsql/epgsql/issues/179#issuecomment-457319416

@jarlopez interesting. That sounds like Postgres would always re-send the last message, regardless of the client's handling. In that case do you end up having to track the last read message in the consuming application to not process it twice?

**Comment by jarlopez** [comment]
@jeffutter Yep, I track the last committed LSN in the Elixir process state, and I deduplicate events at the consumer of the logical decoding worker.
